gh-91401: Add a failsafe way to disable vfork.

gpshead opened the pull request:
Just in case there is ever an issue with `_posixsubprocess`'s use of
`vfork()` due to the complexity of using it properly and potential
directions that Linux platforms where it defaults to on could take, this
adds a failsafe so that users can disable its use entirely by setting
a global flag.

No known reason to disable it exists. But it'd be a shame to encounter
one and not be able to use CPython without patching and rebuilding it.

See the linked issue for some discussion on reasoning.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

gpshead committed:
gh-91401: Add a failsafe way to disable vfork.

Just in case there is ever an issue with _posixsubprocess's use of
vfork() due to the complexity of using it properly and potential
directions that Linux platforms where it defaults to on could take, this
adds a failsafe so that users can disable its use entirely by setting
a global flag.

No known reason to disable it exists. But it'd be a shame to encounter
one and not be able to use CPython without patching and rebuilding it.

See the linked issue for some discussion on reasoning.

vstinner commented on Doc/library/subprocess.rst:
```diff
@@ -1529,3 +1529,29 @@ runtime):
+
+If you ever encounter a presumed highly-unusual situation where you need to
+prevent ``vfork()`` from being used by Python, you can set the
+:attr:`subprocess.disable_vfork_reason` attribute to a non-empty string.
```
It may help to add an example:

```
subprocess.disable_vfork_reason = "no vfork"
```

gpshead replied:
done. simplified.

vstinner reviewed (commented):
Why not a simple boolean flag?

Can subprocess disable vfork using disable_vfork_reason if vfork is not available? You can expose a simple private flag in _posixsubprocess saying if vfork is available. Something similar to _testcapi.WITH_PYMALLOC or _thread.TIMEOUT_MAX.

gpshead committed:
Switch the attr name to _USE_VFORK for consistency

No double negatives. Also documents _USE_POSIX_SPAWN.

gpshead committed:
Add a comment, default to True.

Rather than re-encoding our platform selection here (configure.ac
and runtime logic in _posixsubprocess already does that).

gpshead committed:
NEWS entry.

vstinner commented on Doc/library/subprocess.rst:
```diff
@@ -1529,3 +1529,39 @@ runtime):
+code.
+
+.. versionadded:: 3.8 ``_USE_POSIX_SPAWN``
+.. versionadded:: 3.11 ``_USE_VFORK``
```
I never tried this syntax, I don't know if it's properly rendered. You will see :-)

vstinner reviewed (approved):
LGTM. I just have a doubt about the doc format, but you will see when it will be rendered on docs.python.org.

gpshead committed:
Add a unittest for _USE_VFORK.
